### PR TITLE
(FACT-1136) Move Ruby API to Leatherman

### PR DESCRIPTION
**Pull request by Magisus**
Changed some exceptions back to warnings when messages were not facter specific

### CODE REVIEW

**Comment by MikaelSmith** [comment]
Oh, Windows is assuming auto defaults to value rather than ref.

File: ruby/src/api.cc [hunk 1]
```diff
@@ -136,11 +136,7 @@ namespace leatherman { namespace ruby {
         } else {
             LOG_INFO("ruby was already loaded.");
         }
-        try {
-            return unique_ptr<api>(new api(move(library)));
-        } catch (lth_lib::missing_import_exception& ex) {
-            throw ex;
-        }
+        return unique_ptr<api>(new api(move(library)));
```
Comment (MikaelSmith) [inline ruby/src/api.cc]: I'd lean towards dropping the overhead of unique_ptr. We can just do `return api(move(library));`, assign to `static api instance = create();`, and `return instance;`.

**Comment by MikaelSmith** [comment]
And these can be dealt with as separate commits, just prefix with (FACT-1136).

File: ruby/src/api.cc (hunk 1 above)
Comment (Magisus) [inline ruby/src/api.cc]: It's complaining that this is using a deleted constructor. Do we want to un-delete said constructor, now that we intend to use it this way? Or is there some other workaround?

Comment (MikaelSmith) [inline ruby/src/api.cc]: Use `auto &` instead of `auto`. It thinks you're asking it to create a copy.

Comment (Magisus) [inline ruby/src/api.cc]: The error is coming from line 139 though... no auto involved. Though that did fix line 125.

Comment (MikaelSmith) [inline ruby/src/api.cc]: Hmm, did you update the function return value signature?

Comment (Magisus) [inline ruby/src/api.cc]: Getting a different error now, `non-const lvalue reference to type 'leatherman::ruby::api' cannot bind to a temporary of type 'leatherman::ruby::api'`. Which I think I understand, but am not sure how to fix correctly.

**Comment by MikaelSmith** [comment]
```
From 1bab0167a9499caffd0000f19acdfca10e9b370e Mon Sep 17 00:00:00 2001
From: Michael Smith <michael.smith@puppetlabs.com>
Date: Tue, 28 Jul 2015 12:06:08 -0700
Subject: [PATCH 3/3] fix

---
 ruby/tests/api-test.cc | 145 +++++++++++++++++++++++--------------------------
 1 file changed, 69 insertions(+), 76 deletions(-)

diff --git a/ruby/tests/api-test.cc b/ruby/tests/api-test.cc
index 1277cf8..a65c452 100644
--- a/ruby/tests/api-test.cc
+++ b/ruby/tests/api-test.cc
@@ -6,152 +6,146 @@ using namespace leatherman::ruby;
 
 TEST_CASE("api::eval", "[ruby-api]") {
     SECTION("can load api and evaluate ruby code") {
-        auto ruby = api::instance();
-        REQUIRE(ruby);
-        ruby->initialize();
-        REQUIRE(ruby->initialized());
+        auto &ruby = api::instance();
+        ruby.initialize();
+        REQUIRE(ruby.initialized());
 
-        REQUIRE(ruby->get_load_path().size() > 0);
+        REQUIRE(ruby.get_load_path().size() > 0);
 
-        REQUIRE(ruby->to_string(ruby->eval("'foo'")) == "foo");
+        REQUIRE(ruby.to_string(ruby.eval("'foo'")) == "foo");
     }
 }
 
 TEST_CASE("api::is_*", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
 
     SECTION("can correctly identify nil values") {
-        REQUIRE(ruby->is_nil(ruby->nil_value()));
-        REQUIRE_FALSE(ruby->is_nil(ruby->true_value()));
+        REQUIRE(ruby.is_nil(ruby.nil_value()));
+        REQUIRE_FALSE(ruby.is_nil(ruby.true_value()));
     }
 
     SECTION("can correctly identify true and false values") {
-        REQUIRE(ruby->is_true(ruby->true_value()));
-        REQUIRE_FALSE(ruby->is_true(ruby->false_value()));
-        REQUIRE(ruby->is_false(ruby->false_value()));
-        REQUIRE_FALSE(ruby->is_false(ruby->true_value()));
+        REQUIRE(ruby.is_true(ruby.true_value()));
+        REQUIRE_FALSE(ruby.is_true(ruby.false_value()));
+        REQUIRE(ruby.is_false(ruby.false_value()));
+        REQUIRE_FALSE(ruby.is_false(ruby.true_value()));
     }
 
     SECTION("can correctly identify strings") {
-        REQUIRE(ruby->is_string(ruby->utf8_value("'I'm a string'")));
-        REQUIRE_FALSE(ruby->is_string(ruby->true_value()));
+        REQUIRE(ruby.is_string(ruby.utf8_value("'I'm a string'")));
+        REQUIRE_FALSE(ruby.is_string(ruby.true_value()));
     }
 
     SECTION("can correctly identify symbols") {
-        REQUIRE(ruby->is_symbol(ruby->to_symbol("mysymbol")));
-        REQUIRE_FALSE(ruby->is_symbol(ruby->false_value()));
+        REQUIRE(ruby.is_symbol(ruby.to_symbol("mysymbol")));
+        REQUIRE_FALSE(ruby.is_symbol(ruby.false_value()));
     }
 
     SECTION("can correctly identify numbers") {
-        REQUIRE(ruby->is_float(ruby->eval("1.5")));
-        REQUIRE_FALSE(ruby->is_float(ruby->utf8_value("foo")));
+        REQUIRE(ruby.is_float(ruby.eval("1.5")));
+        REQUIRE_FALSE(ruby.is_float(ruby.utf8_value("foo")));
 
-        REQUIRE(ruby->is_fixednum(ruby->eval("2")));
-        REQUIRE_FALSE(ruby->is_fixednum(ruby->eval("1.5")));
+        REQUIRE(ruby.is_fixednum(ruby.eval("2")));
+        REQUIRE_FALSE(ruby.is_fixednum(ruby.eval("1.5")));
     }
 
     SECTION("can correctly identify hashes") {
-        REQUIRE(ruby->is_hash(ruby->eval("{ 'red' => 2 }")));
-        REQUIRE_FALSE(ruby->is_hash(ruby->utf8_value("foo")));
+        REQUIRE(ruby.is_hash(ruby.eval("{ 'red' => 2 }")));
+        REQUIRE_FALSE(ruby.is_hash(ruby.utf8_value("foo")));
     }
 
     SECTION("can correctly identify type") {
-        REQUIRE(ruby->is_a(ruby->eval("1"), ruby->eval("Integer")));
-        REQUIRE_FALSE(ruby->is_a(ruby->eval("'1'"), ruby->eval("Integer")));
+        REQUIRE(ruby.is_a(ruby.eval("1"), ruby.eval("Integer")));
+        REQUIRE_FALSE(ruby.is_a(ruby.eval("'1'"), ruby.eval("Integer")));
     }
 
     SECTION("can correctly identify arrays") {
-        REQUIRE(ruby->is_array(ruby->eval("[1, 2, 3]")));
-        REQUIRE_FALSE(ruby->is_array(ruby->false_value()));
+        REQUIRE(ruby.is_array(ruby.eval("[1, 2, 3]")));
+        REQUIRE_FALSE(ruby.is_array(ruby.false_value()));
     }
 }
 
 TEST_CASE("api::equals", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
     SECTION("can correctly test boolean values for equality") {
-        REQUIRE(ruby->equals(ruby->true_value(), ruby->true_value()));
-        REQUIRE_FALSE(ruby->equals(ruby->true_value(), ruby->false_value()));
+        REQUIRE(ruby.equals(ruby.true_value(), ruby.true_value()));
+        REQUIRE_FALSE(ruby.equals(ruby.true_value(), ruby.false_value()));
     }
 
     SECTION("can correctly test strings for equality") {
-        REQUIRE(ruby->equals(ruby->utf8_value("foo"), ruby->utf8_value("foo")));
-        REQUIRE_FALSE(ruby->equals(ruby->utf8_value("foo"), ruby->utf8_value("bar")));
+        REQUIRE(ruby.equals(ruby.utf8_value("foo"), ruby.utf8_value("foo")));
+        REQUIRE_FALSE(ruby.equals(ruby.utf8_value("foo"), ruby.utf8_value("bar")));
     }
 
     SECTION("can correctly test numbers for equality") {
-        REQUIRE(ruby->equals(ruby->eval("1"), ruby->eval("1")));
-        REQUIRE_FALSE(ruby->equals(ruby->eval("1"), ruby->eval("3")));
-        REQUIRE(ruby->equals(ruby->eval("1.5"), ruby->eval("1.5")));
-        REQUIRE_FALSE(ruby->equals(ruby->eval("1.5"), ruby->eval("1")));
+        REQUIRE(ruby.equals(ruby.eval("1"), ruby.eval("1")));
+        REQUIRE_FALSE(ruby.equals(ruby.eval("1"), ruby.eval("3")));
+        REQUIRE(ruby.equals(ruby.eval("1.5"), ruby.eval("1.5")));
+        REQUIRE_FALSE(ruby.equals(ruby.eval("1.5"), ruby.eval("1")));
     }
 
     SECTION("can correctly test Ruby hashes for equality") {
-        REQUIRE(ruby->equals(ruby->eval("{ 'red' => 'blue' }"), ruby->eval("{ 'red' => 'blue' }")));
-        REQUIRE_FALSE(ruby->equals(ruby->eval("{ 'red' => 'blue' }"), ruby->eval("{ 'red' => 'green' }")));
+        REQUIRE(ruby.equals(ruby.eval("{ 'red' => 'blue' }"), ruby.eval("{ 'red' => 'blue' }")));
+        REQUIRE_FALSE(ruby.equals(ruby.eval("{ 'red' => 'blue' }"), ruby.eval("{ 'red' => 'green' }")));
     }
 
     SECTION("can correctly test symbols for equality") {
-        REQUIRE(ruby->equals(ruby->to_symbol("mysymbol"), ruby->eval(":mysymbol")));
-        REQUIRE_FALSE(ruby->equals(ruby->to_symbol("mysymbol"), ruby->to_symbol("notmysymbol")));
+        REQUIRE(ruby.equals(ruby.to_symbol("mysymbol"), ruby.eval(":mysymbol")));
+        REQUIRE_FALSE(ruby.equals(ruby.to_symbol("mysymbol"), ruby.to_symbol("notmysymbol")));
     }
 }
 
 TEST_CASE("api::case_equals", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
 
     SECTION("can detect class membership") {
-        REQUIRE(ruby->case_equals(ruby->eval("Integer"), ruby->eval("1")));
-        REQUIRE_FALSE(ruby->case_equals(ruby->eval("String"), ruby->eval("4")));
+        REQUIRE(ruby.case_equals(ruby.eval("Integer"), ruby.eval("1")));
+        REQUIRE_FALSE(ruby.case_equals(ruby.eval("String"), ruby.eval("4")));
     }
 }
 
 VALUE test_func(VALUE self) {
-    auto ruby = api::instance();
-    ruby->initialize();
-    return ruby->utf8_value("test function");
+    auto &ruby = api::instance();
+    ruby.initialize();
+    return ruby.utf8_value("test function");
 }
 
 TEST_CASE("api::rb_define_singleton_method", "[ruby-api]") {
     SECTION("can define a new module with a new method") {
-        auto ruby = api::instance();
-        REQUIRE(ruby);
-        ruby->initialize();
-        REQUIRE(ruby->initialized());
+        auto &ruby = api::instance();
+        ruby.initialize();
+        REQUIRE(ruby.initialized());
 
-        auto module = ruby->rb_define_module("Test");
+        auto module = ruby.rb_define_module("Test");
         REQUIRE(module);
-        ruby->rb_define_singleton_method(module, "test_func", RUBY_METHOD_FUNC(test_func), 0);
-        REQUIRE(ruby->to_string(ruby->eval("Test.test_func")) == "test function");
+        ruby.rb_define_singleton_method(module, "test_func", RUBY_METHOD_FUNC(test_func), 0);
+        REQUIRE(ruby.to_string(ruby.eval("Test.test_func")) == "test function");
     }
 }
 
 TEST_CASE("api::exception_to_string", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
 
     SECTION("can print exception details") {
         try {
-            ruby->eval("raise 'test_exception'");
+            ruby.eval("raise 'test_exception'");
         } catch (runtime_error exc) {
             REQUIRE(string(exc.what()) == "test_exception");
         }
     }
 
     SECTION("can print exception details with stack trace") {
-        ruby->include_stack_trace(true);
+        ruby.include_stack_trace(true);
         try {
-            ruby->eval("raise 'test_exception'");
+            ruby.eval("raise 'test_exception'");
         } catch (runtime_error exc) {
             REQUIRE(string(exc.what()).find("backtrace") != string::npos);
         }
@@ -159,14 +153,13 @@ TEST_CASE("api::exception_to_string", "[ruby-api]") {
 }
 
 TEST_CASE("api::lookup", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
 
     SECTION("can find module by name") {
-        auto foo_module = ruby->rb_define_module("Foo");
-        ruby->rb_define_module_under(foo_module, "Bar");
-        REQUIRE(ruby->to_string(ruby->lookup({ "Foo", "Bar" })) == "Foo::Bar");
+        auto foo_module = ruby.rb_define_module("Foo");
+        ruby.rb_define_module_under(foo_module, "Bar");
+        REQUIRE(ruby.to_string(ruby.lookup({ "Foo", "Bar" })) == "Foo::Bar");
     }
 }
-- 
2.4.6

```

File: ruby/src/api.cc (hunk 1 above)
Comment (MikaelSmith) [inline ruby/src/api.cc]: Ah, move operator is also deleted. I guess forgot it then, we can use a unique_ptr internally.

**Comment by MikaelSmith** [comment]
Full patch would be something like
```
From 72ea5053de5e02979363d83a2ac3ec6573b0abf5 Mon Sep 17 00:00:00 2001
From: Michael Smith <michael.smith@puppetlabs.com>
Date: Tue, 28 Jul 2015 12:06:08 -0700
Subject: [PATCH] fix

---
 ruby/inc/leatherman/ruby/api.hpp |   2 +-
 ruby/src/api.cc                  |   8 +--
 ruby/tests/api-test.cc           | 145 +++++++++++++++++++--------------------
 3 files changed, 74 insertions(+), 81 deletions(-)

diff --git a/ruby/inc/leatherman/ruby/api.hpp b/ruby/inc/leatherman/ruby/api.hpp
index b48d879..53b3fff 100644
--- a/ruby/inc/leatherman/ruby/api.hpp
+++ b/ruby/inc/leatherman/ruby/api.hpp
@@ -640,7 +640,7 @@ namespace leatherman {  namespace ruby {
         void* (* const ruby_options)(int, char**);
         int (* const ruby_cleanup)(volatile int);
 
-        static std::unique_ptr<api> create();
+        static leatherman::dynamic_library::dynamic_library create();
         static leatherman::dynamic_library::dynamic_library find_library();
         static leatherman::dynamic_library::dynamic_library find_loaded_library();
         static VALUE callback_thunk(VALUE parameter);
diff --git a/ruby/src/api.cc b/ruby/src/api.cc
index a102c25..3479c67 100644
--- a/ruby/src/api.cc
+++ b/ruby/src/api.cc
@@ -122,11 +122,11 @@ namespace leatherman { namespace ruby {
 
     api& api::instance()
     {
-        static unique_ptr<api> instance = create();
-        return *instance.get();
+        static api instance{move(create())};
+        return instance;
     }
 
-    unique_ptr<api> api::create()
+    lth_lib::dynamic_library api::create()
     {
         lth_lib::dynamic_library library = find_library();
         if (!library.loaded()) {
@@ -136,7 +136,7 @@ namespace leatherman { namespace ruby {
         } else {
             LOG_INFO("ruby was already loaded.");
         }
-        return unique_ptr<api>(new api(move(library)));
+        return library;
     }
 
     void api::initialize()
diff --git a/ruby/tests/api-test.cc b/ruby/tests/api-test.cc
index 1277cf8..a65c452 100644
--- a/ruby/tests/api-test.cc
+++ b/ruby/tests/api-test.cc
@@ -6,152 +6,146 @@ using namespace leatherman::ruby;
 
 TEST_CASE("api::eval", "[ruby-api]") {
     SECTION("can load api and evaluate ruby code") {
-        auto ruby = api::instance();
-        REQUIRE(ruby);
-        ruby->initialize();
-        REQUIRE(ruby->initialized());
+        auto &ruby = api::instance();
+        ruby.initialize();
+        REQUIRE(ruby.initialized());
 
-        REQUIRE(ruby->get_load_path().size() > 0);
+        REQUIRE(ruby.get_load_path().size() > 0);
 
-        REQUIRE(ruby->to_string(ruby->eval("'foo'")) == "foo");
+        REQUIRE(ruby.to_string(ruby.eval("'foo'")) == "foo");
     }
 }
 
 TEST_CASE("api::is_*", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
 
     SECTION("can correctly identify nil values") {
-        REQUIRE(ruby->is_nil(ruby->nil_value()));
-        REQUIRE_FALSE(ruby->is_nil(ruby->true_value()));
+        REQUIRE(ruby.is_nil(ruby.nil_value()));
+        REQUIRE_FALSE(ruby.is_nil(ruby.true_value()));
     }
 
     SECTION("can correctly identify true and false values") {
-        REQUIRE(ruby->is_true(ruby->true_value()));
-        REQUIRE_FALSE(ruby->is_true(ruby->false_value()));
-        REQUIRE(ruby->is_false(ruby->false_value()));
-        REQUIRE_FALSE(ruby->is_false(ruby->true_value()));
+        REQUIRE(ruby.is_true(ruby.true_value()));
+        REQUIRE_FALSE(ruby.is_true(ruby.false_value()));
+        REQUIRE(ruby.is_false(ruby.false_value()));
+        REQUIRE_FALSE(ruby.is_false(ruby.true_value()));
     }
 
     SECTION("can correctly identify strings") {
-        REQUIRE(ruby->is_string(ruby->utf8_value("'I'm a string'")));
-        REQUIRE_FALSE(ruby->is_string(ruby->true_value()));
+        REQUIRE(ruby.is_string(ruby.utf8_value("'I'm a string'")));
+        REQUIRE_FALSE(ruby.is_string(ruby.true_value()));
     }
 
     SECTION("can correctly identify symbols") {
-        REQUIRE(ruby->is_symbol(ruby->to_symbol("mysymbol")));
-        REQUIRE_FALSE(ruby->is_symbol(ruby->false_value()));
+        REQUIRE(ruby.is_symbol(ruby.to_symbol("mysymbol")));
+        REQUIRE_FALSE(ruby.is_symbol(ruby.false_value()));
     }
 
     SECTION("can correctly identify numbers") {
-        REQUIRE(ruby->is_float(ruby->eval("1.5")));
-        REQUIRE_FALSE(ruby->is_float(ruby->utf8_value("foo")));
+        REQUIRE(ruby.is_float(ruby.eval("1.5")));
+        REQUIRE_FALSE(ruby.is_float(ruby.utf8_value("foo")));
 
-        REQUIRE(ruby->is_fixednum(ruby->eval("2")));
-        REQUIRE_FALSE(ruby->is_fixednum(ruby->eval("1.5")));
+        REQUIRE(ruby.is_fixednum(ruby.eval("2")));
+        REQUIRE_FALSE(ruby.is_fixednum(ruby.eval("1.5")));
     }
 
     SECTION("can correctly identify hashes") {
-        REQUIRE(ruby->is_hash(ruby->eval("{ 'red' => 2 }")));
-        REQUIRE_FALSE(ruby->is_hash(ruby->utf8_value("foo")));
+        REQUIRE(ruby.is_hash(ruby.eval("{ 'red' => 2 }")));
+        REQUIRE_FALSE(ruby.is_hash(ruby.utf8_value("foo")));
     }
 
     SECTION("can correctly identify type") {
-        REQUIRE(ruby->is_a(ruby->eval("1"), ruby->eval("Integer")));
-        REQUIRE_FALSE(ruby->is_a(ruby->eval("'1'"), ruby->eval("Integer")));
+        REQUIRE(ruby.is_a(ruby.eval("1"), ruby.eval("Integer")));
+        REQUIRE_FALSE(ruby.is_a(ruby.eval("'1'"), ruby.eval("Integer")));
     }
 
     SECTION("can correctly identify arrays") {
-        REQUIRE(ruby->is_array(ruby->eval("[1, 2, 3]")));
-        REQUIRE_FALSE(ruby->is_array(ruby->false_value()));
+        REQUIRE(ruby.is_array(ruby.eval("[1, 2, 3]")));
+        REQUIRE_FALSE(ruby.is_array(ruby.false_value()));
     }
 }
 
 TEST_CASE("api::equals", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
     SECTION("can correctly test boolean values for equality") {
-        REQUIRE(ruby->equals(ruby->true_value(), ruby->true_value()));
-        REQUIRE_FALSE(ruby->equals(ruby->true_value(), ruby->false_value()));
+        REQUIRE(ruby.equals(ruby.true_value(), ruby.true_value()));
+        REQUIRE_FALSE(ruby.equals(ruby.true_value(), ruby.false_value()));
     }
 
     SECTION("can correctly test strings for equality") {
-        REQUIRE(ruby->equals(ruby->utf8_value("foo"), ruby->utf8_value("foo")));
-        REQUIRE_FALSE(ruby->equals(ruby->utf8_value("foo"), ruby->utf8_value("bar")));
+        REQUIRE(ruby.equals(ruby.utf8_value("foo"), ruby.utf8_value("foo")));
+        REQUIRE_FALSE(ruby.equals(ruby.utf8_value("foo"), ruby.utf8_value("bar")));
     }
 
     SECTION("can correctly test numbers for equality") {
-        REQUIRE(ruby->equals(ruby->eval("1"), ruby->eval("1")));
-        REQUIRE_FALSE(ruby->equals(ruby->eval("1"), ruby->eval("3")));
-        REQUIRE(ruby->equals(ruby->eval("1.5"), ruby->eval("1.5")));
-        REQUIRE_FALSE(ruby->equals(ruby->eval("1.5"), ruby->eval("1")));
+        REQUIRE(ruby.equals(ruby.eval("1"), ruby.eval("1")));
+        REQUIRE_FALSE(ruby.equals(ruby.eval("1"), ruby.eval("3")));
+        REQUIRE(ruby.equals(ruby.eval("1.5"), ruby.eval("1.5")));
+        REQUIRE_FALSE(ruby.equals(ruby.eval("1.5"), ruby.eval("1")));
     }
 
     SECTION("can correctly test Ruby hashes for equality") {
-        REQUIRE(ruby->equals(ruby->eval("{ 'red' => 'blue' }"), ruby->eval("{ 'red' => 'blue' }")));
-        REQUIRE_FALSE(ruby->equals(ruby->eval("{ 'red' => 'blue' }"), ruby->eval("{ 'red' => 'green' }")));
+        REQUIRE(ruby.equals(ruby.eval("{ 'red' => 'blue' }"), ruby.eval("{ 'red' => 'blue' }")));
+        REQUIRE_FALSE(ruby.equals(ruby.eval("{ 'red' => 'blue' }"), ruby.eval("{ 'red' => 'green' }")));
     }
 
     SECTION("can correctly test symbols for equality") {
-        REQUIRE(ruby->equals(ruby->to_symbol("mysymbol"), ruby->eval(":mysymbol")));
-        REQUIRE_FALSE(ruby->equals(ruby->to_symbol("mysymbol"), ruby->to_symbol("notmysymbol")));
+        REQUIRE(ruby.equals(ruby.to_symbol("mysymbol"), ruby.eval(":mysymbol")));
+        REQUIRE_FALSE(ruby.equals(ruby.to_symbol("mysymbol"), ruby.to_symbol("notmysymbol")));
     }
 }
 
 TEST_CASE("api::case_equals", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
 
     SECTION("can detect class membership") {
-        REQUIRE(ruby->case_equals(ruby->eval("Integer"), ruby->eval("1")));
-        REQUIRE_FALSE(ruby->case_equals(ruby->eval("String"), ruby->eval("4")));
+        REQUIRE(ruby.case_equals(ruby.eval("Integer"), ruby.eval("1")));
+        REQUIRE_FALSE(ruby.case_equals(ruby.eval("String"), ruby.eval("4")));
     }
 }
 
 VALUE test_func(VALUE self) {
-    auto ruby = api::instance();
-    ruby->initialize();
-    return ruby->utf8_value("test function");
+    auto &ruby = api::instance();
+    ruby.initialize();
+    return ruby.utf8_value("test function");
 }
 
 TEST_CASE("api::rb_define_singleton_method", "[ruby-api]") {
     SECTION("can define a new module with a new method") {
-        auto ruby = api::instance();
-        REQUIRE(ruby);
-        ruby->initialize();
-        REQUIRE(ruby->initialized());
+        auto &ruby = api::instance();
+        ruby.initialize();
+        REQUIRE(ruby.initialized());
 
-        auto module = ruby->rb_define_module("Test");
+        auto module = ruby.rb_define_module("Test");
         REQUIRE(module);
-        ruby->rb_define_singleton_method(module, "test_func", RUBY_METHOD_FUNC(test_func), 0);
-        REQUIRE(ruby->to_string(ruby->eval("Test.test_func")) == "test function");
+        ruby.rb_define_singleton_method(module, "test_func", RUBY_METHOD_FUNC(test_func), 0);
+        REQUIRE(ruby.to_string(ruby.eval("Test.test_func")) == "test function");
     }
 }
 
 TEST_CASE("api::exception_to_string", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
 
     SECTION("can print exception details") {
         try {
-            ruby->eval("raise 'test_exception'");
+            ruby.eval("raise 'test_exception'");
         } catch (runtime_error exc) {
             REQUIRE(string(exc.what()) == "test_exception");
         }
     }
 
     SECTION("can print exception details with stack trace") {
-        ruby->include_stack_trace(true);
+        ruby.include_stack_trace(true);
         try {
-            ruby->eval("raise 'test_exception'");
+            ruby.eval("raise 'test_exception'");
         } catch (runtime_error exc) {
             REQUIRE(string(exc.what()).find("backtrace") != string::npos);
         }
@@ -159,14 +153,13 @@ TEST_CASE("api::exception_to_string", "[ruby-api]") {
 }
 
 TEST_CASE("api::lookup", "[ruby-api]") {
-    auto ruby = api::instance();
-    REQUIRE(ruby);
-    ruby->initialize();
-    REQUIRE(ruby->initialized());
+    auto &ruby = api::instance();
+    ruby.initialize();
+    REQUIRE(ruby.initialized());
 
     SECTION("can find module by name") {
-        auto foo_module = ruby->rb_define_module("Foo");
-        ruby->rb_define_module_under(foo_module, "Bar");
-        REQUIRE(ruby->to_string(ruby->lookup({ "Foo", "Bar" })) == "Foo::Bar");
+        auto foo_module = ruby.rb_define_module("Foo");
+        ruby.rb_define_module_under(foo_module, "Bar");
+        REQUIRE(ruby.to_string(ruby.lookup({ "Foo", "Bar" })) == "Foo::Bar");
     }
 }
-- 
2.4.6
```

**Comment by MikaelSmith** [comment]
:+1:

File: ruby/src/api.cc [hunk 2]
```diff
@@ -136,11 +136,7 @@ namespace leatherman { namespace ruby {
         } else {
             LOG_INFO("ruby was already loaded.");
         }
-        try {
-            return unique_ptr<api>(new api(move(library)));
-        } catch (lth_lib::missing_import_exception& ex) {
```
Comment (peterhuene) [inline ruby/src/api.cc]: Where did this catch go?

Comment (peterhuene) [inline ruby/src/api.cc]: Erm, never mind, for some reason it was being rethrown (incorrectly too).  Did this return `nullptr` at one point?

File: ruby/src/api.cc [hunk 3]
```diff
@@ -474,7 +469,7 @@ namespace leatherman { namespace ruby {
                                                                   "break file if File.exist? file;"
                                                                   "false end)" });
         if (!success) {
-            throw execution_failure_exception("ruby failed to run", output, none);
+            LOG_WARNING("ruby failed to run", output, none);
```
Comment (peterhuene) [inline ruby/src/api.cc]: This should raise a format exception from boost.  There's no `%1%` or `%2%` in the message.

File: ruby/src/api.cc [hunk 4]
```diff
@@ -452,7 +447,7 @@ namespace leatherman { namespace ruby {
             if (library.load(value)) {
                 return library;
             } else {
-                throw library_not_loaded_exception((boost::format("ruby library \"%1%\" could not be loaded.") % value).str());
+                LOG_WARNING((boost::format("ruby library \"%1%\" could not be loaded.") % value).str());
```
Comment (peterhuene) [inline ruby/src/api.cc]: The formatting could have been performed by the macro.

File: ruby/src/api.cc [hunk 5]
```diff
@@ -441,8 +437,7 @@ namespace leatherman { namespace ruby {
             if (library.load(ruby_lib_location)) {
                 return library;
             }
-            throw library_not_loaded_exception(
-                    (boost::format("preferred ruby library \"%1%\" could not be loaded.") % ruby_lib_location).str());
+            LOG_WARNING((boost::format("preferred ruby library \"%1%\" could not be loaded.") % ruby_lib_location).str());
```
Comment (peterhuene) [inline ruby/src/api.cc]: The formatting could have been performed by the macro.

File: ruby/src/api.cc (hunk 2 above)
Comment (MikaelSmith) [inline ruby/src/api.cc]: Yes, returned nullptr before. I suggested using exceptions to pass useful debugging messages back to the caller.

File: ruby/src/api.cc (hunk 5 above)
Comment (MikaelSmith) [inline ruby/src/api.cc]: Very true. @Magisus looks like a few minor fixups we should still do.

**Pull request merged**